lib/Controller/PathController.php

```diff
@@ -4,89 +4,319 @@
+				\OC\Files\Filesystem::filesize($filename));
+		}
+
+		if (\OC\Files\Filesystem::isReadable($filename)) {
+			self::sendHeaders($filename, $name, $rangeArray);
+		} elseif (!\OC\Files\Filesystem::file_exists($filename)) {
+			http_response_code(404);
+			$tmpl = new OC_Template('', '404', 'guest');
+			$tmpl->printPage();
+			exit();
+		} else {
+			http_response_code(403);
+			die('403 Forbidden');
+		}
+		if (isset($params['head']) && $params['head']) {
+			return;
+		}
+		if (!empty($rangeArray)) {
+			try {
+				if (count($rangeArray) == 1) {
+					$view->readfilePart($filename, $rangeArray[0]['from'], $rangeArray[0]['to']);
+				}
+				else {
+					// check if file is seekable (if not throw UnseekableException)
+					// we have to check it before body contents
+					$view->readfilePart($filename, $rangeArray[0]['size'], $rangeArray[0]['size']);
+					$type = \OC::$server->getMimeTypeDetector()->getSecureMimeType(\OC\Files\Filesystem::getMimeType($filename));
+					foreach ($rangeArray as $range) {
+						echo "\r\n--".self::getBoundary()."\r\n".
+							"Content-type: ".$type."\r\n".
+							"Content-range: bytes ".$range['from']."-".$range['to']."/".$range['size']."\r\n\r\n";
+						$view->readfilePart($filename, $range['from'], $range['to']);
+					}
+					echo "\r\n--".self::getBoundary()."--\r\n";
+				}
+			} catch (\OCP\Files\UnseekableException $ex) {
+				// file is unseekable
+				header_remove('Accept-Ranges');
+				header_remove('Content-Range');
+				http_response_code(200);
+				self::sendHeaders($filename, $name, array());
+				$view->readfile($filename);
+			}
+		}
+		else {
+			$output = $view->readfile($filename);
+		}
+	}
+
+	/**
+	 * @param string $dir
+	 * @param $files
+	 * @param integer $getType
+	 * @param View $view
+	 * @param string $filename
+	 */
+	private static function unlockAllTheFiles($dir, $files, $getType, $view, $filename) {
+		if ($getType === self::FILE) {
+			$view->unlockFile($filename, ILockingProvider::LOCK_SHARED);
+		}
+		if ($getType === self::ZIP_FILES) {
+			foreach ($files as $file) {
+				$file = $dir . '/' . $file;
+				$view->unlockFile($file, ILockingProvider::LOCK_SHARED);
+			}
+		}
+		if ($getType === self::ZIP_DIR) {
+			$file = $dir . '/' . $files;
+			$view->unlockFile($file, ILockingProvider::LOCK_SHARED);
+		}
+	}
+
+	/**
+	 * return the content of a file or return a zip file containing multiple files
+	 *
+	 * @param string $dir
+	 * @param string $files ; separated list of files to download
+	 * @param array $params ; 'head' boolean to only send header of the request ; 'range' http range header
+	 */
+	public static function get($dir, $files, $params = null) {
+		$view = \OC\Files\Filesystem::getView();
+		$getType = self::FILE;
+		try {
+			$filename = $dir . '/' . $files;
+			if (!$view->is_dir($filename)) {
+				self::getSingleFile($view, $dir, $files, is_null($params) ? array() : $params);
+				return;
+			}
+			set_time_limit($executionTime);
```

Comment:
undefined var $executionTime

Reply:
I just commited fixes to the pull request, please review. This point should be fixed as well.